ports/nrf/hal/hal_nvmc.c

```diff
@@ -0,0 +1,204 @@
+    return hal_nvmc_operation_state == HAL_NVMC_STATE_SUCCESS;
+}
+
+bool hal_nvmc_erase_page(uint32_t pageaddr) {
+    operation_init();
+    uint32_t result = sd_flash_page_erase(pageaddr / HAL_NVMC_PAGESIZE);
+    return operation_complete(result);
+}
+
+bool hal_nvmc_write_words(uint32_t *dest, const uint32_t *buf, size_t len) {
+    operation_init();
+    uint32_t result = sd_flash_write(dest, buf, len);
+    return operation_complete(result);
+}
+
+bool hal_nvmc_write_byte(byte *dest_in, byte b) {
+    uint32_t dest = (uint32_t)dest_in;
+    uint32_t dest_aligned = dest & ~3;
+
+    // Value to write - leave all bits that should not change at 0xff.
+    uint32_t value = 0xffffff00 | b;
+
+    // Rotate bits in value to an aligned position.
+    for (int i = dest - dest_aligned; i; i--) {
+        value = (value << 8) | 0xff;
+    }
+
+    operation_init();
+    uint32_t result = sd_flash_write((uint32_t*)dest_aligned, &value, 1);
+    return operation_complete(result);
+}
+
+#else // BLUETOOTH_SD
+
+bool hal_nvmc_erase_page(uint32_t pageaddr) {
+    // Configure NVMC to erase a page.
+    NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Een;
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // Set the page to erase
+    NRF_NVMC->ERASEPAGE = pageaddr;
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // Switch back to read-only.
+    NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Ren;
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // Operation succeeded.
+    return true;
+}
+
+bool hal_nvmc_write_words(uint32_t *dest, const uint32_t *buf, size_t len) {
+    // Note that we're writing 32-bit integers, not bytes. Thus the 'real'
+    // length of the buffer is len*4.
+
+    // Configure NVMC so that writes are allowed (anywhere).
+    NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Wen;
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // Write all integers to flash.
+    for (int i=0; i<len; i++) {
+        dest[i] = buf[i];
+        while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+    }
+
+    // Switch back to read-only.
+    NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Ren;
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // Operation succeeded.
+    return true;
+}
+
+bool hal_nvmc_write_byte(byte *dest_in, byte b) {
+    // This code can probably be optimized.
+
+    // Configure NVMC so that writes are allowed (anywhere).
+    NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Wen;
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
+
+    // According to the nRF51 RM (chapter 6), only word writes to
+    // word-aligned addresses are allowed.
+    // https://www.nordicsemi.com/eng/nordic/Products/nRF51822/nRF51-RM/62725
+    uint32_t dest = (uint32_t)dest_in;
+    uint32_t dest_aligned = dest & ~3;
+
+    // Value to write - leave all bits that should not change at 0xff.
+    uint32_t value = 0xffffff00 | b;
+
+    // Rotate bits in value to an aligned position.
+    for (int i = dest - dest_aligned; i; i--) {
```

Comment:
is this extra i; necessary?

Reply:
You mean the middle argument to the for loop (the condition)? Yes, otherwise it won't stop looping. I'll change it to `i > 0` or something similar for clarity.